zeek/Shaping-Zeek-NDJSON.md

```diff
@@ -0,0 +1,210 @@
+was originally present inside Zeek. This detail can be restored using a Zed
+shaper, such as the reference example [`shaper.zed`](shaper.zed)
+that can be found in this directory of the repository.
+
+A full description of all that's possible with shapers is beyond the scope of
+this doc. However, this example for shaping Zeek NDJSON is quite simple and
+is described below.
+
+# Reference Shaper Contents
+
+The example `shaper.zed` may seem large, but ultimately it follows a
+fairly simple pattern that repeats across the many [Zeek log types](https://docs.zeek.org/en/master/script-reference/log-files.html).
+
+## Leading Type Definitions
+
+The top three lines define types that are referenced further below in the main
+portion of the Zed.
+
+```
+type port=uint16;
+type zenum=string;
+type conn_id={orig_h:ip,orig_p:port,resp_h:ip,resp_p:port};
+```
+The `port` and `zenum` types are described further in the [Zed/Zeek Data Type Compatibility](Data-Type-Compatibility.md)
+doc. The `conn_id` type will just save us from having to repeat these fields
+individually in the many Zeek record types that contain an embedded `id`
+record.
+
+## Type Definitions Per Zeek Log `_path`
+
+The bulk of this Zed shaper consists of detailed per-field data type
+definitions for each record in the default set of NDJSON logs output by Zeek.
+These type definitions reference the types we defined above, such as `port`
+and `conn_id`. The syntax for defining primitive and complex types follows the
+relevant sections of the [ZSON Format](../docs/formats/zson.md#3-the-zson-format)
+specification.
+
+```
+...
+type conn={_path:string,ts:time,uid:bstring,id:conn_id,proto:zenum,service:bstring,duration:duration,orig_bytes:uint64,resp_bytes:uint64,conn_state:bstring,local_orig:bool,local_resp:bool,missed_bytes:uint64,history:bstring,orig_pkts:uint64,orig_ip_bytes:uint64,resp_pkts:uint64,resp_ip_bytes:uint64,tunnel_parents:|[bstring]|,_write_ts:time};
+type dce_rpc={_path:string,ts:time,uid:bstring,id:conn_id,rtt:duration,named_pipe:bstring,endpoint:bstring,operation:bstring,_write_ts:time};
+...
+```
+
+> **Note:** See [the role of `_path`](Reading-Zeek-Log-Formats.md#the-role-of-_path)
+> doc for important details if you're using Zeek's built-in [ASCII logger](https://docs.zeek.org/en/current/scripts/base/frameworks/logging/writers/ascii.zeek.html)
+> to generate NDJSON rather than the [JSON Streaming Logs](https://github.com/corelight/json-streaming-logs) package.
+
+## Mapping From `_path` Values to Types
+
+The next section is just simple mapping from the string values typically found
+in the Zeek `_path` field to the name of one of the types we defined above.
+
+```
+const schemas = |{
+  "broker": broker,
+  "capture_loss": capture_loss,
+  "cluster": cluster,
+  "config": config,
+  "conn": conn,
+  "dce_rpc": dce_rpc,
+...
+```
+
+## Zed Pipeline
+
+The Zed ends with a pipeline that stitches together everything we've defined
+so far.
```

Comment:
As above, I think this reads more smoothly if Zed is used as an adjective as it is elsewhere.
```suggestion
The Zed shaper ends with a pipeline that stitches together everything we've defined
so far.
```

Reply:
@nwt: Yes, happy to make this change. In many ways, I'm still getting the hang of (and am somewhat conflicted by) the many uses of "Zed". I'm often telling myself that not all Zed is "queries". Part of me would like to be able to call it "Zed code", but that feels a little too programming-language-y, and I'm not sure if I want to make that a habit. This is one place where I somewhat miss "ZQL" because that felt like three-letter shorthand for "ya know... the stuff in the language" without triggering any obsession (for me, at least) over the "Q" standing for "query". In this specific case, you're correct that it seems fully appropriate to explicitly call it a "Zed shaper", but if we had shorthand to just refer to any bit of code (query, shaper, other ETL, etc.) like we used to have with "ZQL", this is a place where I would have done it just in the interest of brevity. But it's a small difference regardless.

Anyway, I'll keep at it until it feels natural.